File: python/tests/test_black_scholes.py
"""Black-Scholes analytics: known values, parity, Greeks vs FD, edge cases."""

from __future__ import annotations

import math

import pytest

from localvol import (
    Market,
    bs_delta,
    bs_gamma,
    bs_price,
    bs_vega,
    implied_vol,
)


def test_known_atm_value():
    # Classic textbook number: S=K=100, r=5%, q=0, sigma=20%, T=1.
    assert bs_price(100, 100, 0.05, 0.0, 0.2, 1.0, True) == pytest.approx(10.450584, abs=1e-6)


def test_put_call_parity_grid():
    """Property-style: parity C - P = S e^{-qT} - K e^{-rT} across a grid."""
    for s in [80.0, 100.0, 123.4]:
        for k in [60.0, 100.0, 140.0]:
            for r, q in [(0.05, 0.0), (0.01, 0.03), (-0.005, 0.0)]:
                for t in [0.1, 1.0, 3.0]:
                    c = bs_price(s, k, r, q, 0.25, t, True)
                    p = bs_price(s, k, r, q, 0.25, t, False)
                    lhs = c - p
                    rhs = s * math.exp(-q * t) - k * math.exp(-r * t)
                    assert lhs == pytest.approx(rhs, abs=1e-10)


def test_garman_kohlhagen_via_market_fx():
    mkt = Market.fx(1.20, rd=0.03, rf=0.01)
    # GK call == BS call with q = rf
    assert bs_price(mkt.spot, 1.25, mkt.rate, mkt.dividend, 0.10, 0.5, True) == pytest.approx(
        bs_price(1.20, 1.25, 0.03, 0.01, 0.10, 0.5, True)
    )
    assert mkt.forward(0.5) == pytest.approx(1.20 * math.exp(0.02 * 0.5))


def test_edge_cases():
    assert bs_price(100, 90, 0.05, 0.0, 0.2, 0.0, True) == 10.0  # expiry -> intrinsic
    assert bs_price(100, 110, 0.05, 0.0, 0.2, 0.0, True) == 0.0
    # sigma=0 -> discounted forward intrinsic
    f = 100 * math.exp(0.05)
    assert bs_price(100, 90, 0.05, 0.0, 0.0, 1.0, True) == pytest.approx(math.exp(-0.05) * (f - 90))
    assert bs_price(100, 0.0, 0.03, 0.01, 0.2, 1.0, True) == pytest.approx(100 * math.exp(-0.01))
    assert bs_price(100, 0.0, 0.03, 0.01, 0.2, 1.0, False) == 0.0
    # deep ITM/OTM: values pinned to asymptotics
    assert bs_price(100, 1e6, 0.0, 0.0, 0.2, 1.0, True) == pytest.approx(0.0, abs=1e-12)
    deep_itm = bs_price(100, 1.0, 0.0, 0.0, 0.2, 1.0, True)
    assert deep_itm == pytest.approx(99.0, abs=1e-8)


def test_input_validation():
    with pytest.raises(ValueError):
        bs_price(-1.0, 100, 0.0, 0.0, 0.2, 1.0)
    with pytest.raises(ValueError):
        bs_price(100, -5.0, 0.0, 0.0, 0.2, 1.0)
    with pytest.raises(ValueError):
        bs_price(100, 100, 0.0, 0.0, -0.2, 1.0)
    with pytest.raises(ValueError):
        bs_price(100, 100, 0.0, 0.0, 0.2, -1.0)
    with pytest.raises(ValueError):
        bs_price(float("nan"), 100, 0.0, 0.0, 0.2, 1.0)
    with pytest.raises(ValueError):
        Market(-1.0, 0.0, 0.0)


def test_greeks_vs_finite_differences():
    s, k, r, q, sig, t = 105.0, 100.0, 0.03, 0.01, 0.22, 0.75
    eps = 1e-4
    d_fd = (bs_price(s + eps, k, r, q, sig, t) - bs_price(s - eps, k, r, q, sig, t)) / (2 * eps)
    g_fd = (bs_price(s + eps, k, r, q, sig, t) - 2 * bs_price(s, k, r, q, sig, t) + bs_price(s - eps, k, r, q, sig, t)) / eps**2
    v_fd = (bs_price(s, k, r, q, sig + eps, t) - bs_price(s, k, r, q, sig - eps, t)) / (2 * eps)
    assert bs_delta(s, k, r, q, sig, t) == pytest.approx(d_fd, abs=1e-6)
    assert bs_gamma(s, k, r, q, sig, t) == pytest.approx(g_fd, abs=1e-5)
    assert bs_vega(s, k, r, q, sig, t) == pytest.approx(v_fd, abs=1e-4)


def test_implied_vol_round_trip():
    for sig in [0.05, 0.2, 0.8]:
        for k in [70.0, 100.0, 130.0]:
            price = bs_price(100, k, 0.02, 0.01, sig, 1.5, True)
            iv = implied_vol(price, 100, k, 0.02, 0.01, 1.5, True)
            assert iv == pytest.approx(sig, abs=1e-8)


def test_implied_vol_rejects_arbitrage_and_bad_inputs():
    with pytest.raises(ValueError, match="bounds"):
        implied_vol(1000.0, 100, 100, 0.0, 0.0, 1.0, True)  # above S
    with pytest.raises(ValueError, match="bounds"):
        implied_vol(-1.0, 100, 100, 0.0, 0.0, 1.0, True)  # below intrinsic
    with pytest.raises(ValueError):
        implied_vol(5.0, 100, 100, 0.0, 0.0, 0.0, True)  # T = 0


def test_implied_vol_rejects_bad_bracket_and_iterations():
    """MIN-7: non-default lo/hi/iterations are validated."""
    price = bs_price(100, 100, 0.02, 0.01, 0.2, 1.0, True)
    with pytest.raises(ValueError, match="bracket"):
        implied_vol(price, 100, 100, 0.02, 0.01, 1.0, True, lo=5.0, hi=1e-9)
    with pytest.raises(ValueError, match="bracket"):
        implied_vol(price, 100, 100, 0.02, 0.01, 1.0, True, lo=0.0, hi=1.0)
    with pytest.raises(ValueError, match="bracket"):
        implied_vol(price, 100, 100, 0.02, 0.01, 1.0, True, lo=1e-9, hi=float("inf"))
    with pytest.raises(ValueError, match="iterations"):
        implied_vol(price, 100, 100, 0.02, 0.01, 1.0, True, iterations=0)
    with pytest.raises(ValueError, match="price must be finite"):
        implied_vol(float("nan"), 100, 100, 0.02, 0.01, 1.0, True)
    # the default contract: exactly 100 halvings of [1e-9, 5]
    assert implied_vol(price, 100, 100, 0.02, 0.01, 1.0, True) == pytest.approx(0.2, abs=1e-10)


def test_market_log_forward_validation_and_value():
    """MIN-6: log_forward validates expiry in every port."""
    mkt = Market(100.0, 0.03, 0.01)
    assert mkt.log_forward(2.0) == pytest.approx(math.log(100.0) + 0.02 * 2.0, abs=1e-15)
    assert mkt.log_forward(0.0) == math.log(100.0)
    with pytest.raises(ValueError):
        mkt.log_forward(-1.0)
    with pytest.raises(ValueError):
        mkt.log_forward(float("nan"))
    with pytest.raises(ValueError):
        mkt.forward(float("inf"))
    with pytest.raises(ValueError):
        Market(100.0, float("nan"), 0.0)
    with pytest.raises(ValueError):
        Market(100.0, 0.0, float("inf"))

File: cpp/tests/test_black_scholes.cpp
// Black-Scholes analytics: known values, put-call parity over a grid,
// finite-difference Greeks, edge cases, bisection implied vol.

#include <gtest/gtest.h>

#include <cmath>
#include <limits>
#include <stdexcept>

#include "localvol/black_scholes.hpp"
#include "localvol/market.hpp"

using localvol::bs_delta;
using localvol::bs_gamma;
using localvol::bs_price;
using localvol::bs_vega;
using localvol::implied_vol;
using localvol::norm_cdf;

TEST(BlackScholes, KnownAtmValue) {
    // Classic textbook value: S=K=100, r=5%, q=0, sigma=20%, T=1.
    EXPECT_NEAR(bs_price(100, 100, 0.05, 0.0, 0.2, 1.0, true), 10.450583572185565, 1e-9);
}

TEST(BlackScholes, NormCdfSymmetryAndTails) {
    EXPECT_DOUBLE_EQ(norm_cdf(0.0), 0.5);
    EXPECT_NEAR(norm_cdf(1.959963984540054), 0.975, 1e-9);
    EXPECT_NEAR(norm_cdf(-8.0) + norm_cdf(8.0), 1.0, 1e-15);
    EXPECT_GT(norm_cdf(-37.0), 0.0);  // erfc keeps the far tail alive
}

TEST(BlackScholes, PutCallParityGrid) {
    // Property-style loop: C - P = S e^{-qT} - K e^{-rT} across a grid.
    const double s = 100.0, r = 0.03, q = 0.015;
    for (double strike : {60.0, 80.0, 100.0, 120.0, 150.0}) {
        for (double t : {0.1, 0.5, 1.0, 2.0}) {
            for (double sig : {0.1, 0.25, 0.5}) {
                const double c = bs_price(s, strike, r, q, sig, t, true);
                const double p = bs_price(s, strike, r, q, sig, t, false);
                const double rhs = s * std::exp(-q * t) - strike * std::exp(-r * t);
                EXPECT_NEAR(c - p, rhs, 1e-10)
                    << "K=" << strike << " T=" << t << " sig=" << sig;
            }
        }
    }
}

TEST(BlackScholes, GarmanKohlhagenFxPut) {
    // FX put via r=rd, q=rf must equal the same formula by construction and
    // satisfy parity with the GK forward.
    const double c = bs_price(1.10, 1.05, 0.03, 0.01, 0.10, 0.5, true);
    const double p = bs_price(1.10, 1.05, 0.03, 0.01, 0.10, 0.5, false);
    const double rhs = 1.10 * std::exp(-0.01 * 0.5) - 1.05 * std::exp(-0.03 * 0.5);
    EXPECT_NEAR(c - p, rhs, 1e-12);
    EXPECT_GT(p, 0.0);
}

TEST(BlackScholes, EdgeCases) {
    EXPECT_DOUBLE_EQ(bs_price(100, 90, 0.05, 0.0, 0.2, 0.0, true), 10.0);   // T=0 intrinsic
    EXPECT_DOUBLE_EQ(bs_price(100, 110, 0.05, 0.0, 0.2, 0.0, true), 0.0);
    const double f = 100.0 * std::exp(0.05);
    EXPECT_NEAR(bs_price(100, 90, 0.05, 0.0, 0.0, 1.0, true),
                std::exp(-0.05) * (f - 90.0), 1e-12);                        // sigma=0
    EXPECT_NEAR(bs_price(100, 0.0, 0.05, 0.02, 0.2, 1.0, true),
                100.0 * std::exp(-0.02), 1e-12);                             // K=0 call
    EXPECT_DOUBLE_EQ(bs_price(100, 0.0, 0.05, 0.02, 0.2, 1.0, false), 0.0);  // K=0 put
}

TEST(BlackScholes, GreeksMatchFiniteDifferences) {
    const double s = 100, k = 105, r = 0.02, q = 0.01, sig = 0.25, t = 0.75;
    const double eps = 1e-4;
    const double dfd =
        (bs_price(s + eps, k, r, q, sig, t, true) - bs_price(s - eps, k, r, q, sig, t, true)) /
        (2 * eps);
    EXPECT_NEAR(bs_delta(s, k, r, q, sig, t, true), dfd, 1e-6);
    const double gfd = (bs_price(s + eps, k, r, q, sig, t, true) -
                        2 * bs_price(s, k, r, q, sig, t, true) +
                        bs_price(s - eps, k, r, q, sig, t, true)) /
                       (eps * eps);
    EXPECT_NEAR(bs_gamma(s, k, r, q, sig, t), gfd, 1e-5);
    const double veps = 1e-5;
    const double vfd = (bs_price(s, k, r, q, sig + veps, t, true) -
                        bs_price(s, k, r, q, sig - veps, t, true)) /
                       (2 * veps);
    EXPECT_NEAR(bs_vega(s, k, r, q, sig, t), vfd, 1e-5);
    // Put delta from parity.
    EXPECT_NEAR(bs_delta(s, k, r, q, sig, t, false),
                bs_delta(s, k, r, q, sig, t, true) - std::exp(-q * t), 1e-12);
}

TEST(BlackScholes, ImpliedVolRoundTrip) {
    for (double sig : {0.08, 0.2, 0.45, 1.2}) {
        for (double strike : {80.0, 100.0, 125.0}) {
            const double price = bs_price(100, strike, 0.03, 0.01, sig, 0.9, true);
            const double iv = implied_vol(price, 100, strike, 0.03, 0.01, 0.9, true);
            EXPECT_NEAR(iv, sig, 1e-8) << "sig=" << sig << " K=" << strike;
        }
    }
    // Puts too.
    const double pp = bs_price(100, 95, 0.02, 0.0, 0.3, 0.5, false);
    EXPECT_NEAR(implied_vol(pp, 100, 95, 0.02, 0.0, 0.5, false), 0.3, 1e-8);
}

TEST(BlackScholes, ImpliedVolRejectsOutOfBoundsPrices) {
    // Below the sigma=0 lower bound.
    EXPECT_THROW(implied_vol(0.0, 100, 50, 0.05, 0.0, 1.0, true), std::invalid_argument);
    // Above the discounted-spot upper bound.
    EXPECT_THROW(implied_vol(101.0, 100, 100, 0.0, 0.0, 1.0, true), std::invalid_argument);
    EXPECT_THROW(implied_vol(5.0, 100, 100, 0.05, 0.0, 0.0, true), std::invalid_argument);
    EXPECT_THROW(implied_vol(5.0, 100, 0.0, 0.05, 0.0, 1.0, true), std::invalid_argument);
}

TEST(BlackScholes, RejectsInvalidInputs) {
    const double nan = std::nan("");
    EXPECT_THROW(bs_price(-1, 100, 0.0, 0.0, 0.2, 1.0), std::invalid_argument);
    EXPECT_THROW(bs_price(100, -1, 0.0, 0.0, 0.2, 1.0), std::invalid_argument);
    EXPECT_THROW(bs_price(100, 100, 0.0, 0.0, -0.2, 1.0), std::invalid_argument);
    EXPECT_THROW(bs_price(100, 100, 0.0, 0.0, 0.2, -1.0), std::invalid_argument);
    EXPECT_THROW(bs_price(100, 100, nan, 0.0, 0.2, 1.0), std::invalid_argument);
    EXPECT_THROW(bs_gamma(100, 100, 0.0, 0.0, 0.0, 1.0), std::invalid_argument);
    EXPECT_THROW(bs_vega(100, 100, 0.0, 0.0, 0.2, 0.0), std::invalid_argument);
}

TEST(BlackScholes, ImpliedVolRejectsBadBracketAndIterations) {
    // MIN-7: non-default lo/hi/iterations are validated.
    const double price = localvol::bs_price(100, 100, 0.02, 0.01, 0.2, 1.0, true);
    EXPECT_THROW(localvol::implied_vol(price, 100, 100, 0.02, 0.01, 1.0, true, 5.0, 1e-9),
                 std::invalid_argument);
    EXPECT_THROW(localvol::implied_vol(price, 100, 100, 0.02, 0.01, 1.0, true, 0.0, 1.0),
                 std::invalid_argument);
    EXPECT_THROW(localvol::implied_vol(price, 100, 100, 0.02, 0.01, 1.0, true, 1e-9,
                                       std::numeric_limits<double>::infinity()),
                 std::invalid_argument);
    EXPECT_THROW(localvol::implied_vol(price, 100, 100, 0.02, 0.01, 1.0, true, 1e-9, 5.0, 0),
                 std::invalid_argument);
    EXPECT_THROW(localvol::implied_vol(std::nan(""), 100, 100, 0.02, 0.01, 1.0, true),
                 std::invalid_argument);
    EXPECT_NEAR(localvol::implied_vol(price, 100, 100, 0.02, 0.01, 1.0, true), 0.2, 1e-10);
}

TEST(BlackScholes, MarketLogForwardValidationAndValue) {
    // MIN-6: log_forward validates expiry in every port.
    const localvol::Market mkt(100.0, 0.03, 0.01);
    EXPECT_NEAR(mkt.log_forward(2.0), std::log(100.0) + 0.02 * 2.0, 1e-15);
    EXPECT_DOUBLE_EQ(mkt.log_forward(0.0), std::log(100.0));
    EXPECT_THROW(mkt.log_forward(-1.0), std::invalid_argument);
    EXPECT_THROW(mkt.log_forward(std::nan("")), std::invalid_argument);
    EXPECT_THROW(mkt.forward(std::numeric_limits<double>::infinity()), std::invalid_argument);
    EXPECT_THROW(localvol::Market(100.0, std::nan(""), 0.0), std::invalid_argument);
    EXPECT_THROW(localvol::Market(100.0, 0.0, std::numeric_limits<double>::infinity()),
                 std::invalid_argument);
}
